pkg/cli/cli.go

```diff
@@ -1362,27 +1362,55 @@ func runConnectionWebSocketAgentCommand(flags *ConnectFlags, c *command, args []
 		return err
 	}
 	adbPort := l.Addr().(*net.TCPAddr).Port
-	go func() {
-		if err := opts.ADBServerProxy.Connect(adbPort); err != nil {
+	adbErrCh := make(chan error)
+	go func(adbErrCh chan error) {
+		err := opts.ADBServerProxy.Connect(adbPort)
+		if err != nil {
 			c.PrintErrf("Failed to connect ADB to device: %v\n", err)
 		}
-	}()
+		adbErrCh <- err
+	}(adbErrCh)
 
-	tcpListener, ok := l.(*net.TCPListener)
-	if !ok {
-		c.PrintErrf("Listener is not a TCPListener\n")
-		return errors.New("Listener is not a TCPListener")
-	}
-	tcpListener.SetDeadline(time.Now().Add(time.Duration(adbListenTimeoutSec) * time.Second))
+	listenResultCh := make(chan net.Conn)
+	listenErrCh := make(chan error)
+	go func(listenResultCh chan net.Conn, listenErrCh chan error) {
+		tcpListener, ok := l.(*net.TCPListener)
+		if !ok {
+			listenErrCh <- errors.New("listener is not a TCPListener")
+			return
+		}
+		tcpListener.SetDeadline(time.Now().Add(time.Duration(adbListenTimeoutSec) * time.Second))
 
-	tcpConn, err := l.Accept()
-	if err != nil {
-		if ne, ok := err.(net.Error); ok && ne.Timeout() {
-			c.PrintErrf("Waiting connection from ADB server timed out: ")
-		} else {
-			c.PrintErrf("Failed to accept ADB socket: %v", err)
+		tcpConn, err := l.Accept()
+		if err != nil {
+			if ne, ok := err.(net.Error); ok && ne.Timeout() {
+				c.PrintErrf("Waiting connection from ADB server timed out: ")
+			} else {
+				c.PrintErrf("Failed to accept ADB socket: %v", err)
+			}
+			listenErrCh <- err
+			return
+		}
+		listenResultCh <- tcpConn
+	}(listenResultCh, listenErrCh)
+
+	var tcpConn net.Conn
+	for tcpConn == nil {
+		select {
+		case err := <-adbErrCh:
+			if err != nil {
+				return err
```

Comment:
nit: An early return here would cause the other channels to not be read from, which in turn would block the routine trying to send on that channel. The same happens with the return in line 1406. That's unlikely to cause major issues but it would be cleaner to make sure that both routines finish before returning from this function. That can be done by receiving from `adbErrCh` first and then a `select` with the other two channels (a `select` won't be needed if you follow my previous advice).
Or, just don't use a routine for accepting on the listener, that has the same effect as receiving from those channels before receiving from `adbErrCh`.

Reply:
Removed go routine for listener, and modified codes to read from adbErrCh channel to avoid the situation mentioned.